# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [ ]:
# review in recording

In [1]:
def add_two(x):
    return x + 2

print (add_two(4))

6


#### Saving to a Variable

In [2]:
# written as a Lambda function
# Self invoking function - Meaning it immediately calls itself
add_three = lambda x: x + 3

print(add_three)

<function <lambda> at 0x7fe21a5c5040>


#### Multiple Inputs

In [3]:
# Multiples inputs with no variables
(lambda x,y,z: x * y / z)(6,4,2)

12.0

In [4]:
# Multiple Inputs with a variabale attached

mult_test = lambda x, y, z: x * y / z
mult_test(8, 6, 4)

12.0

#### Passing a Lambda into a Function

In [6]:
def multiply(f, num):
    """
    f expects a function that accepts one argument
    """
    return (f)
    num -= 50 
    return f(num)


multiply(lambda x: (x + 5)/10, -100)

100

In [ ]:
abs(-100)

#### Returning a Lambda from a Function

In [ ]:
# 

#### If Statements within Lambdas

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [7]:
f = lambda num: num**3

f(10)

1000

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

#### Using Lambda's with Map

In [ ]:
my_list = [1,2,3,4,5,6,7,9,4,3,2,]
my_list

In [ ]:
list(map(lambda x: x**2/2, my_list))

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [10]:
a_list = [10, 35, 66, 3, 6, 34, 22, 13, 87]

list(map(lambda n: n*2-1, a_list))

[19, 69, 131, 5, 11, 67, 43, 25, 173]

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [11]:
help(filter)

Help on class filter in module builtins:

class filter(object)
 |  filter(function or None, iterable) --> filter object
 |  
 |  Return an iterator yielding those items of iterable for which function(item)
 |  is true. If function is None, return the items that are true.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



In [12]:
def is_even(num):
    if num % 2 == 0:
        return True
    else:
        return False
    
list_b = [1,2,3,4,5,6,7,8,9,10]

list(filter(is_even, list_b))

[2, 4, 6, 8, 10]

#### Using Lambda's with Filter()

In [16]:
def a_names(name):
    return name[0] == 'A'
    
a_names('Andrew')

True

In [13]:
def a_names(name):
    if name[0] == 'A':
        return True
    else:
        return False

In [14]:
names = ['Bob', 'Andrew', 'Annie', 'Max', 'Evan', 'Alex']

a_names = list(filter(lambda name: True if name[0] == 'A' else False, names))
print(a_names)

['Andrew', 'Annie', 'Alex']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [22]:
import statistics

mean_of_list = statistics.mean(a_list)

def is_greater_than_mean(num):
    return num > mean_of_list


is_greater_than_mean(92)

greater_than = []
for number in a_list:
    if is_greater_than_mean(number):
        greater_than.append(number)

print(greater_than)

[92, 77, 81, 82, 72, 68, 70, 84, 91, 83, 94]


In [28]:
import statistics

a_list = [41, 48, 92, 77, 5, 81, 82, 72, 68, 57, 70, 84, 56, 91, 38, 5, 29, 24, 83, 94]
print(a_list)

a = statistics.mean(a_list)

print("Mean is:", a)


        

[41, 48, 92, 77, 5, 81, 82, 72, 68, 57, 70, 84, 56, 91, 38, 5, 29, 24, 83, 94]
Mean is: 59.85


In [29]:
list(filter(lambda x: x > statistics.mean(a_list), a_list))

[92, 77, 81, 82, 72, 68, 70, 84, 91, 83, 94]

In [26]:
list(filter(lambda x: x > sum(a_list)/len(a_list), a_list))

[92, 77, 81, 82, 72, 68, 70, 84, 91, 83, 94]

In [27]:
list(map(lambda x: x > sum(a_list)/len(a_list), a_list))

[False,
 False,
 True,
 True,
 False,
 True,
 True,
 True,
 True,
 False,
 True,
 True,
 False,
 True,
 False,
 False,
 False,
 False,
 True,
 True]

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [30]:
# reduce(function, iterable)
from functools import reduce

help(reduce)

Help on built-in function reduce in module _functools:

reduce(...)
    reduce(function, sequence[, initial]) -> value
    
    Apply a function of two arguments cumulatively to the items of a sequence,
    from left to right, so as to reduce the sequence to a single value.
    For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) calculates
    ((((1+2)+3)+4)+5).  If initial is present, it is placed before the items
    of the sequence in the calculation, and serves as a default when the
    sequence is empty.



In [36]:
list1 = [1,2,3,4,5]

def mult_nums(num1, num2):
    return num1 * num2

reduce(mult_nums, list1)

120

In [34]:
# more pythonic in reference to zen of python
total = 1
for num in list1:
    total *= num
print(total)

120


#### Using Lambda's with Reduce()

In [32]:
list1 = [1,2,3,4,5]

reduce(lambda x, y: x*y, list1)

120

#### In-Class Example #4 <br>
<p>Use reduce to find the max of a list .</p>

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [ ]:
# review in recording

In [41]:
def another_func(a_num):
    print('Another func is running')
    return f"The number is {a_num}" 

def my_func(num):
    print(f"My number is {num}")
    num += 1
    return another_func(num)

print(f"The return value of my_func when I pass in 10 is {my_func(10)}")

My number is 10
Another func is running
The return value of my_func when I pass in 10 is The number is 11


#### Writing a Factorial Function

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [42]:
def my_range(stop, start, step = 2):
    while start < stop:
        yield start
        start += step
    
    
for i in my_range(50, 3):
    print(i)

3
5
7
9
11
13
15
17
19
21
23
25
27
29
31
33
35
37
39
41
43
45
47
49


#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [173]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

filter_places = list(filter(lambda x: x.strip(), places))
print(filter_places)

['Argentina', 'San Diego', 'Boston', 'New York']


In [172]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

filter_places = filter(lambda x: len(x) > 2, places)

without_places = list(filter_places)
print (without_places)

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [86]:
# lambda
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

author.sort(key=lambda name: name.split(" ")[-1].lower())
print(author)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


In [90]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]
x = sorted(author, key=lambda name: name.split(" ")[-1].lower())
print(x)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [99]:
# map and lambda
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

In [100]:
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

def celsius(place, temp):
      return place, (9/5)*int(temp) + 32

location = ['Nashua', "Boston", "Los Angelos", "Miami"]
temperature = [32, 12, 44, 29]

c_f = list(map(celsius,location, temperature))
print(c_f)

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


In [101]:
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

location = ['Nashua', "Boston", "Los Angelos", "Miami"]
temperature = [32, 12, 44, 29]

ce_fa = list(map(lambda place, temp: (place, (9/5)*int(temp) + 32), location, temperature))
print(ce_fa)

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [180]:
def fibonacci(n):
    a, b, counter = 1, 1, 0
    while True:
        if (counter > n): 
            return
        yield a 
        a, b = b, a + b
        print(f"Iteration {counter}: {x}")
        counter += 1    
        
f = fibonacci(5)                  
my_list =[]
for x in f:
    my_list.append(x)

Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8
